### DISCUSSION TITLE
Add support for Remix Vite

### DISCUSSION THREAD

**Discussion by alexdotpink** [nrwl/nx#22613] [Feature Requests]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description
Adding support for Remix with Vite in the @nx/remix plugin.

## Motivation
I believe using Remix with Vite allows for a better developer experience.

## Suggested Implementation
Just like for @nx/react, when you generate a @nx/remix plugin you should be able to specify `--bundler=vite`

**Comment by Coly010** [comment]
This is on our roadmap 🙂 

It'll be addressed in due course

**Comment by unrealsolver** [comment]
@Coly010 , Can I (+my team) help you with that?

**Comment by binaryartifex** [comment]
has there been any traction on this yet? has anyone successfully migrated a fresh remix app according to the remix docs using a freshly scaffolded remix app from the nx plugin to the vite build one? ive spent too much time already wrestling with nx's architecture...

**Reply by unrealsolver** [reply]
There is a PR to Nx: https://github.com/nrwl/nx/pull/22757
And there is a @nx/remix fork that I am using for testing: https://github.com/unrealsolver/nx-remix-vite
NOTE: This is work in progress, and there are only executor updates, there are NO generator changes, NO migration from classic to vite. You'll have to manually add a vite.config.ts, make necessary changes to other files. After that you are good to go.

**Comment by hennessyevan** [comment]
Is there any way to make this happen? Even manually? I've setup a vite app and added remix plugin but I get all sorts of weird errors from remix complaining about a missing package.json to some internal nx issues with nx-devkit and things.

**Reply by unrealsolver** [reply]
You can try this fork https://github.com/unrealsolver/nx-remix-vite. Works for my project. But I had hard time accepting it to main repo.  And I didn't updated it to the latest nx version

**Reply by hennessyevan** [reply]
Oh awesome! Do you have an example of what files I'd need to change and what my vite.config should look like? I have a vite.config.ts setup with the nx-vite-plugin and remix-plugin currently but getting all sorts of errors.

_After I installed I got `LoadPluginError: Could not load plugin @nx/remix/plugin` so I downgraded to nx@18 and we're doing okay again._

Now I'm getting `Could not resolve "@angular-devkit/architect"`

**Reply by unrealsolver** [reply]
Basically, just create a normal nx/remix app and follow this guide: https://remix.run/docs/en/main/start/v2
Here is the vite.config.ts:
```typescript
/// <reference types='vitest' />
import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';
import { vitePlugin as remix } from '@remix-run/dev';
import { defineConfig } from 'vite';

export default defineConfig({
  root: __dirname,
  cacheDir: '../../node_modules/.vite/packages/webapp',

  server: {
    port: 4200,
    host: '0.0.0.0',
  },

  preview: {
    port: 8080,
    host: '0.0.0.0',
  },

  plugins: [
    remix({
      ignoredRouteFiles: ['**/*.css'],
    }),
    nxViteTsPaths(),
  ],

  // Uncomment this if you are using workers.
  // worker: {
  //  plugins: [ nxViteTsPaths() ],
  // },

  build: {
    outDir: '../../dist/packages/webapp',
    reportCompressedSize: true,
    commonjsOptions: {
      transformMixedEsModules: true,
    },
  },
});
```
And `remix.config.mjs`
```javascript
import { createWatchPaths } from '@nx/remix';
import { dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

/**
 * @type {import('@remix-run/dev').AppConfig}
 */
export default {
  ignoredRouteFiles: ['**/.*'],
  // appDirectory: "app",
  // assetsBuildDirectory: "public/build",
  // serverBuildPath: "build/index.js",
  // publicPath: "/build/",
  //serverModuleFormat: "cjs",
  serverDependenciesToBundle: [
    /^lodash.*/,
    /^winston.*/,
    /^logform.*/,
    /^@colors*/,
    /^util.*/,
  ],
  postcss: true,
  watchPaths: () => createWatchPaths(__dirname),
};
```

**Comment by wjxalexander** [comment]
An feasible solution is this repo: https://github.com/jaysoo/nx-remix-vite-example
But we need to pass correct [appdirectory](https://remix.run/docs/en/main/file-conventions/vite-config#appdirectory) to the remix vite plugin based on you project

**Reply by ledniy** [reply]
Doesn't look like it's workings solution, I tried to add `appDirectory: "apps/my-remix-app/app"` as well as missing packages to package.json and it still fails 
<img width="1267" alt="image" src="https://github.com/user-attachments/assets/5babbdcd-f078-44c3-95b0-09c543560438">

**Comment by itxch** [comment]
Support seems to be present now :)

As per this PR: https://github.com/nrwl/nx/issues/27884 (and all the other PRs it links to at the bottom).

Example: https://github.com/itxch/nx-remix-vite

The steps I followed:
1. Setup a nx monorepo using `v20`
2. create a remix application using the standard generator
3. Remove all dependancies from the generated `package.json` in the remix app folder
4. Update TSConfig to match a new generated remix app created via `npx create-remix`
5. Create a entry.client and entry.server file by using `npx remix reveal` in my remix app folder
6. Follow the remix to vite migration guide: https://remix.run/docs/en/main/guides/vite#migrating
8. It all works :) all tasks are inferred via the remix plugin



![CleanShot 2024-10-08 at 01 09 59](https://github.com/user-attachments/assets/e85d80e1-518f-4f9c-83be-195f7cdd25a7)

**Reply by afazzdev** [reply]
I don't know why but i can't reproduce this even after i do your guide step by step. can you kindly make it a repo so i can peek?

**Reply by itxch** [reply]
Here ya go https://github.com/itxch/nx-remix-vite

**Reply by superjose** [reply]
@itxch OMG. 

I want to kiss you so hard. 

Remix Vite and a Monorepo structure have **_been very, very, very painful_**. I've wasted [an ungodly number of days](https://hackernoon.com/how-to-create-a-monorepo-with-vite-cloudflare-remix-pnpm-and-turborepo-no-build-step) trying to get something to work. 

I wanted to support Cloudflare Workers/Pages (which I ditched). Not having a build step would allow me to have the remix app pull the Cloudflare-specific dependencies, and then I'd have another nodejs-based app with the NodeJS-specific dependencies. 

This became very tricky. NX didn't have proper support, so I had to wire in [tsx](https://www.npmjs.com/package/tsx) with Vite and ppm workspaces to make it work. After several weeks of failed attempts, I finally solved it. 

I had remix running in Vite, with a monorepo setup.

The problem? 

Yesterday, I hit another wall when I tried importing a class from another package.

I wanted to cry. I've been working on an MVP in which I've taken an ungodly amount of work to have a simple setup, and I was this close to throw the towel and bring everything together.

But your solution has given me hope. I've cloned your repo, installed https://github.com/kiliman/remix-express-vite-plugin, and made a small proof of concept. 

**It's working**

I will take time today to migrate the repository to nx.

Thank you 🙏

**Reply by itxch** [reply]
@superjose  Awesome, really glad its working for you and that its helped

**Reply by diegoromero93** [reply]
Hey @itxch Do you know how to execute the start command ?
`npx nx start remix-app-vite`

I was trying to build and run it like that for production, and I'm encountering this issue

```
> remix-serve /nx-remix-vite/apps/remix-app-vite/build/client

Error: Directory import '/nx-remix-vite/apps/remix-app-vite/build/client' is not supported resolving ES modules imported from /nx-remix-vite/node_modules/@remix-run/serve/dist/cli.js
    at finalizeResolution (node:internal/modules/esm/resolve:258:11)
    at moduleResolve (node:internal/modules/esm/resolve:917:10)
    at defaultResolve (node:internal/modules/esm/resolve:1130:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:396:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:365:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:240:38)
    at ModuleLoader.import (node:internal/modules/esm/loader:328:34)
    at importModuleDynamically (node:internal/modules/cjs/loader:1261:33)
    at importModuleDynamicallyWrapper (node:internal/vm/module:431:21)
    at importModuleDynamicallyCallback (node:internal/modules/esm/utils:207:14)
Warning: command "remix-serve /nx-remix-vite/apps/remix-app-vite/build/client" exited with non-zero status code 
```

**Comment by itxch** [comment]
@Coly010 are there any plans to update the generator/docs for the Remix plugin to use/explain the `vite` support? 
If its not being worked on I'd be happy to raise a PR attempting to do this

**Comment by caprica** [comment]
I'm curious the current status of this vs the endgame, because I'm very interested in using nx with contemporary Remix,.

Is this a fair summary from what I've read here?:

Currently it's only kind-of supported, you need to use the standard nx instructions to generate a Remix project, but after that you need to manually edit multiple configuration files, and then manually migrate the project from classic Remix, and even then there are issues with some imports.

So, out-of-the-box it's not working, but that will be coming, eventually.

**Reply by Coly010** [reply]
It's being worked on currently: https://github.com/nrwl/nx/pull/28555

**Comment by caprica** [comment]
This is included now with 20.0.7 release, and it seems to mostly be OK out-of-the-box, but there are some problems:

1. The README stating the `server` command to run the Remix app does not work, you have to do `nx run my-remix-app:dev` instead.
2. The README showing how to add a library uses the react plugin as an example, maybe it should be the remix library plugin instead.
3. The GitHub action fails in the E2E build.

But what I am most interested in is how to use a custom `vite.config.ts` file in this setup? Does anyone have an example of such a thing that works with this? The config I already have in my Remix project can't just be copied & pasted over, it misses things like root: __dirname, and the server port, but more importantly it misses some configuration to setup the import paths to the other library modules in the monorepo correctly.

I have a minimal pre-made monorepo project here if it's of any interest, but it stops short of having the custom vite config for Remix that I need:
https://github.com/caprica/nxremix

**Reply by caprica** [reply]
> But what I am most interested in is how to use a custom `vite.config.ts` file in this setup? Does anyone have an example of such a thing that works with this?

I found a template file for vite config in the remix plugin repo:

This appears to work, with the imports being resolved correctly, I added the port explicitly since it picked a random port without this:

```typescript
import { vitePlugin as remix } from '@remix-run/dev';
import { defineConfig } from 'vite';
import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

declare module '@remix-run/node' {
    interface Future {
      v3_singleFetch: true;
    }
  }

  export default defineConfig({
    root: __dirname,
    plugins: [
      remix({
        future: {
          v3_fetcherPersist: true,
          v3_relativeSplatPath: true,
          v3_throwAbortReason: true,
          v3_singleFetch: true,
          v3_lazyRouteDiscovery: true,
        },
      }),
      nxViteTsPaths(),
    ],
    server: {
        port: 3000
    }
  });
```
It's not quite 100% though as there are console warnings during startup.

**Reply by caprica** [reply]
> It's not quite 100% though as there are console warnings during startup

To resolve the console warnings during startup, and restore the default server port for Remix applications, I needed to update `vite.config.ts` to add:

```typescript
    server: {
      port: 3000,
      fs: {
        allow: ['..']
      },
    },
```
I am not sure if the `fs.allow` fix is correct, e.g. are there any down-sides or other implications of doing this, I really am not so knowledgeable in Vite configurations.

**Reply by caprica** [reply]
Here is my full latest `vite.config.ts` that *appears* to be fully working:

- default server port for Remix applications
- no errors/warnings on startup or accessing the application
- imports on libraries in the monorepo resolved correctly in the IDE and at runtime
- hot-module-reload working

```typescript
import { vitePlugin as remix } from '@remix-run/dev';
import { defineConfig } from 'vite';
import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

declare module '@remix-run/node' {
    interface Future {
      v3_singleFetch: true;
    }
  }

  export default defineConfig({
    root: __dirname,
    plugins: [
      remix({
        future: {
          v3_fetcherPersist: true,
          v3_relativeSplatPath: true,
          v3_throwAbortReason: true,
          v3_singleFetch: true,
          v3_lazyRouteDiscovery: true,
        },
      }),
      nxViteTsPaths(),
    ],
    server: {
      port: 3000,
      fs: {
        allow: ['..']
      },
    },
  });
```

**Reply by caprica** [reply]
So it all seems to work except the failing e2e action on GitHub.

Using a fresh project, immediately after generation via the nx CLI, and *not* adding a `vite.config.ts` file, the GitHub action fails with:

```
[webkit] › example.spec.ts:3:5 › has title ────────────────────────────────────────────────────

    Error: browserType.launch: 
    ╔══════════════════════════════════════════════════════╗
    ║ Host system is missing dependencies to run browsers. ║
    ║ Missing libraries:                                   ║
    ║     libwoff2dec.so.1.0.2                             ║
    ║     libvpx.so.7                                      ║
    ║     libevent-2.1.so.7                                ║
    ║     libopus.so.0                                     ║
    ║     libgstallocators-1.0.so.0                        ║
    ║     libgstapp-1.0.so.0                               ║
    ║     libgstpbutils-1.0.so.0                           ║
    ║     libgstaudio-1.0.so.0                             ║
    ║     libgstgl-1.0.so.0                                ║
    ║     libgsttag-1.0.so.0                               ║
    ║     libgstvideo-1.0.so.0                             ║
    ║     libgstcodecparsers-1.0.so.0                      ║
    ║     libgstfft-1.0.so.0                               ║
    ║     libharfbuzz-icu.so.0                             ║
    ║     libhyphen.so.0                                   ║
    ║     libmanette-0.2.so.0                              ║
    ║     libflite.so.1                                    ║
    ║     libflite_usenglish.so.1                          ║
    ║     libflite_cmu_grapheme_lang.so.1                  ║
    ║     libflite_cmu_grapheme_lex.so.1                   ║
    ║     libflite_cmu_indic_lang.so.1                     ║
    ║     libflite_cmu_indic_lex.so.1                      ║
    ║     libflite_cmulex.so.1                             ║
    ║     libflite_cmu_time_awb.so.1                       ║
    ║     libflite_cmu_us_awb.so.1                         ║
    ║     libflite_cmu_us_kal16.so.1                       ║
    ║     libflite_cmu_us_kal.so.1                         ║
    ║     libflite_cmu_us_rms.so.1                         ║
    ║     libflite_cmu_us_slt.so.1                         ║
    ║     libGLESv2.so.2                                   ║
    ║     libx264.so                                       ║
    ╚══════════════════════════════════════════════════════╝
```

After I added the vite config file, it fails earlier I guess:

```
> playwright test

[WebServer] The CJS build of Vite's Node API is deprecated. See https://vite.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.

Error: Process from config.webServer was not able to start. Exit code: 1
```

I don't know how to proceed from here.

**Reply by caprica** [reply]
Running the e2e build locally with verbose I get a little more on the error.

I must still be missing something from my custom `vite.config.ts`.

```bash
nx affected -t e2e --verbose
```

```
> nx run my-remix-app-e2e:e2e
> playwright test
[WebServer] Was not able to require.resolve module nx/src/utils/app-root from the following paths: /home/user/workspaces/nx/nxremix/node_modules. This may be expected.
[WebServer] Was not able to require.resolve module nx/src/utils/app-root from the following paths: /home/user/workspaces/nx/nxremix/node_modules. This may be expected.
[WebServer] Error: Directory import '/home/user/workspaces/nx/nxremix/apps/my-remix-app/build/client' is not supported resolving ES modules imported from /home/user/workspaces/nx/nxremix/node_modules/@remix-run/serve/dist/cli.js
[WebServer]     at finalizeResolution (node:internal/modules/esm/resolve:259:11)
[WebServer]     at moduleResolve (node:internal/modules/esm/resolve:933:10)
[WebServer]     at defaultResolve (node:internal/modules/esm/resolve:1169:11)
[WebServer]     at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:383:12)
[WebServer]     at ModuleLoader.resolve (node:internal/modules/esm/loader:352:25)
[WebServer]     at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:227:38)
[WebServer]     at ModuleLoader.import (node:internal/modules/esm/loader:315:34)
[WebServer]     at defaultImportModuleDynamically (node:internal/modules/esm/utils:194:31)
[WebServer]     at importModuleDynamicallyCallback (node:internal/modules/esm/utils:216:12)
[WebServer]     at reimportServer (/home/user/workspaces/nx/nxremix/node_modules/@remix-run/serve/dist/cli.js:86:81)
[WebServer] Warning: command "remix-serve /home/user/workspaces/nx/nxremix/apps/my-remix-app/build/client" exited with non-zero status code

Error: Process from config.webServer was not able to start. Exit code: 1
```

Changing module types in `package.json` just breaks other things.

**Reply by PilatP** [reply]
[Is it](https://github.com/nrwl/nx/pull/28555) what we are waiting for? )

**Reply by caprica** [reply]
No, that PR has already been merged and is available in the current release, and it's what I used entirely in this sub-thread I posted.

With the latest release it does mostly work, but in summary what doesn't work out-of-the-box (for me at least) is this:

- github action fails on e2e test build
- putting your own vite.config.ts in your Remix app mostly works (one of my posts above), but it makes the e2e build fail even earlier

**Reply by caprica** [reply]
The 20.0.8 release with #28738 fixes the original e2e build failure, i.e. installing Playwright --with-deps.

I think my only remaining issue is being unable to have it all working with a `vite.config.ts` in the Remix application folder because of the module loading error I described above when running the e2e target.

>[WebServer] Error: Directory import '/home/user/workspaces/nx/nxremix/apps/my-remix-app/build/client' is not supported resolving ES modules imported from /home/user/workspaces/nx/nxremix/node_modules/@remix-run/serve/dist/cli.js

Is anyone using a custom `vite.config.ts` with this setup?

**Reply by caprica** [reply]
Interesting, I just now upgraded nx to latest, including an updated Remix plugin, at version 20.1.0.

The nx migration deleted my vite.config.ts, but putting it back in there it actually seems to all work now.

I guess I just needed to wait :-)

**Comment by marouanechoukri31** [comment]
Hi, Is there any support to a Remix spa mode app with Nx ?

**Reply by caprica** [reply]
Well, all SPA mode needs is to configure the Remix plugin in the `vite.config.ts` file for the application (by setting `ssr: false`), but you might suffer similar problems as per my sub-thread above when trying to add your own such config file.

**Reply by marouanechoukri31** [reply]
Thank you! I referred to the remix documentation at https://remix.run/docs/en/main/guides/spa-mode, and it worked for me.

**Reply by caprica** [reply]
So, can I ask, you have your own vite.config.ts in the Remix application folder, you enabled SPA mode in that config file, and your *entire* nx project builds, deploys, and runs successfully, including the e2e project with the Playwright tests?

**Reply by marouanechoukri31** [reply]
Yeah, but I haven’t configured the E2E yet. I will let you know when it’s done. Is there a problem with your E2E?

**Reply by caprica** [reply]
Thanks, everything seems to be fixed with the latest version of nx/remix-plugin. I documented my entire experience in this discussion thread :-)
